sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -210,6 +204,39 @@ class NodeHttpClient implements HttpClient {
     return new Promise<http.IncomingMessage>((resolve, reject) => {
       const req = isInsecure ? http.request(options) : https.request(options);
 
+      // The `connectionTimeoutInMs` property defines the number of milliseconds a request can take on stablishing a connection.
+      // If `connectionTimeoutInMs` is not set, the `timeout` property is used for the same purpose.
+      // Defaults to 0, which disables the timeout.
+      const connectionTimeoutInMs = request.connectionTimeoutInMs || request.timeout;
+      let connectionTimeout: NodeJS.Timeout;
+      if (connectionTimeoutInMs > 0) {
+        req.on("socket", (socket) => {
+          // If connecting is true, no connection has been established yet.
+          if (socket.connecting) {
+            connectionTimeout = setTimeout(() => {
+              socket.destroy();
+              abortController.abort();
+            }, connectionTimeoutInMs);
+          }
+          // If connecting is false, the socket is already connected.
+        });
```

Comment:
A few things here:

1. No need to destroy the socket, calling abort on the abortController will call `req.abort` (which we should change to be `req.destroy`) which destroys the socket. Double destroying the socket seems unnecessary. 
2. We should be using `once` instead of `on` so the listeners get cleaned up faster.
3. Personally, I think using the timeout even and letting Node watch the clock on the socket is easier/simpler/cleaner.

```suggestion
        req.setTimeout(connectionTimeoutInMs);
        req.once("timeout", () => {
          abortController.abort();
        });
```

Reply:
This can’t be done since `setTimeout` only gets called once the connection has been stablished.

Reply:
I’ve moved to use `once` and to avoid calling destroy here: https://github.com/Azure/azure-sdk-for-js/pull/16540/commits/9b1a504ab975842452a46082fdd8898eb2bb5aff